zerver/lib/actions.py

```diff
@@ -3459,3 +3460,168 @@ def do_update_user_custom_profile_data(user_profile, data):
+
+    stream_objects = aggregate_dict(
+        input_rows=stream_msgs,
+        lookup_fields=[
+            'stream_id',
+            'topic',
+        ],
+        input_field='message_id',
+        output_field='unread_message_ids',
+    )
+
+    huddle_msgs = [
+        dict(
+            recipient_id=row['message__recipient_id'],
+            message_id=row['message_id'],
+        ) for row in rows
+        if row['message__recipient__type'] == Recipient.HUDDLE]
+
+    huddle_objects = aggregate_dict(
+        input_rows=huddle_msgs,
+        lookup_fields=[
+            'recipient_id',
+        ],
+        input_field='message_id',
+        output_field='unread_message_ids',
+    )
+
+    for huddle in huddle_objects:
+        huddle['user_ids_string'] = huddle_users(huddle['recipient_id'])
+        del huddle['recipient_id']
+
+    result = dict(
+        pms=pm_objects,
+        streams=stream_objects,
+        huddles=huddle_objects,
+    )
+
+    return result
+
+def apply_unread_message_event(state, message):
+    # type: (Dict[str, List[Dict[str, Any]]], Dict[str, Any]) -> None
+    message_id = message['id']
+    if message['type'] == 'stream':
+        message_type = 'stream'
+    elif message['type'] == 'private':
+        others = [
+            recip for recip in message['display_recipient']
+            if recip['id'] != message['sender_id']
+        ]
+        if len(others) <= 1:
+            message_type = 'private'
+        else:
+            message_type = 'huddle'
+
+    if message_type == 'stream':
+        unread_key = 'streams'
+        stream_id = message['stream_id']
+        topic = message['subject']
+
+        my_key = (stream_id, topic)  # type: ignore
+
+        key_func = lambda obj: (obj['stream_id'], obj['topic'])
+        new_obj = dict(
+            stream_id=stream_id,
+            topic=topic,
+            unread_message_ids=[message_id],
+        )
+    elif message_type == 'private':
+        unread_key = 'pms'
+        sender_id = message['sender_id']
+
+        my_key = sender_id  # type: ignore
+        key_func = lambda obj: obj['sender_id']
+        new_obj = dict(
+            sender_id=sender_id,
+            unread_message_ids=[message_id],
+        )
+    else:
+        unread_key = 'huddles'
+        display_recipient = message['display_recipient']
+        user_ids = [obj['id'] for obj in display_recipient]
+        user_ids = sorted(user_ids)
+        my_key = ','.join(str(uid) for uid in user_ids)  # type: ignore
+        key_func = lambda obj: obj['user_ids_string']
+        new_obj = dict(
+            user_ids_string=my_key,
+            unread_message_ids=[message_id],
+        )
+
+    for obj in state[unread_key]:
+        if key_func(obj) == my_key:
+            obj['unread_message_ids'].append(message_id)
+            obj['unread_message_ids'].sort()
+            return
+
+    state[unread_key].append(new_obj)
+    state[unread_key].sort(key=key_func)
```

Comment:
What do you think about putting these functions in `zerver/lib/messages.py`?  That feels like a more natural home.